sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2006,21 +2006,33 @@ class Analyzer(override val catalogManager: CatalogManager)
     override def apply(plan: LogicalPlan): LogicalPlan = {
       val externalFunctionNameSet = new mutable.HashSet[Seq[String]]()
 
-      plan.resolveExpressionsWithPruning(_.containsAnyPattern(UNRESOLVED_FUNCTION)) {
-        case f @ UnresolvedFunction(nameParts, _, _, _, _) =>
-          if (ResolveFunctions.lookupBuiltinOrTempFunction(nameParts).isDefined) {
+      def lookupFunction[T <: TreeNode[_]](
+          f: T,
+          nameParts: Seq[String],
+          lookupBuiltinOrTempFunc: Seq[String] => Option[ExpressionInfo]): T = {
+        if (lookupBuiltinOrTempFunc(nameParts).isDefined) {
+          f
+        } else {
+          val CatalogAndIdentifier(catalog, ident) = expandIdentifier(nameParts)
+          val fullName = normalizeFuncName(catalog.name +: ident.namespace :+ ident.name)
+          if (externalFunctionNameSet.contains(fullName)) {
+            f
+          } else if (catalog.asFunctionCatalog.functionExists(ident)) {
```

Comment:
this checks both scalar and table functions, shall we differentiate them?

One problem is: only the v1 catalog supports table function, so there is no v2 API for it. The logic will be a bit complicatd
```
if (isSessionCatalog) {
  sessionCatalog.tableFunctionExist(...)
} else {
  false
}
```

Reply:
Let me update this PR to only fix the bug in looking up scalar functions. Also, I discovered that the current error message for table functions is more user-friendly: `could not resolve `dummy` to a table-valued function`